src/api/codereader.rs

```diff
@@ -0,0 +1,259 @@
+use byteorder::{BigEndian, ReadBytesExt};
+use std::fs::File;
+use std::io::{Cursor, Read};
+use std::io::{Error, ErrorKind};
+use std::result::Result;
+use std::slice;
+
+use crate::objects::codeholder::CodeHolder;
+use crate::objects::instruction::Instruction;
+use crate::objects::register::{Register, RegisterLocation, RegisterReference};
+
+/// Creates a register instance from 5 bytes
+fn make_register(cur: &mut Cursor<&Vec<u8>>) -> Result<Register, Error> {
+    let reg = cur.read_u32::<BigEndian>()?;
+    let mut locval: u8 = 0;
+    cur.read(slice::from_mut(&mut locval))?;
+
+    let regloc = match locval {
+        0x01 => RegisterLocation::ConstantPool,
+        0x02 => RegisterLocation::Accumulator,
+        0x03 => RegisterLocation::Global,
+        0x04 => RegisterLocation::Local,
+        _ => {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "Invalid RegisterLocation value {} at position {}",
+                    locval,
+                    cur.position() - 1
+                ),
+            ));
+        }
+    };
+
+    Ok(Register(reg, regloc))
+}
+
+/// Creates a register reference
+fn make_rref(cur: &mut Cursor<&Vec<u8>>) -> Result<RegisterReference, Error> {
+    let mut v: u8 = 0;
+    cur.read(slice::from_mut(&mut v))?;
+
+    let rref = match v {
+        0x01 => RegisterReference::AsIs,
+        0x02 => RegisterReference::Dereference,
+        _ => {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!(
+                    "Invalid RegisterReference value {} at position {}",
+                    v,
+                    cur.position() - 1
+                ),
+            ));
+        }
+    };
+
+    Ok(rref)
+}
+
+/// Opens and reads bytecode from a file and parses it into a usable
+/// CodeHolder.
+pub fn read_bytecode_file(path: &str) -> Result<CodeHolder, Error> {
+    let mut file = File::open(path)?;
+    let mut buf = Vec::new(); // has subtype u8
+    file.read_to_end(&mut buf)?;
+
+    let holder = read_bytecode(&buf)?;
+
+    Ok(holder)
+}
+
+/// Parses bytecode contained in a Vec<u8> and returns a usable CodeHolder.
+pub fn read_bytecode(buf: &Vec<u8>) -> Result<CodeHolder, Error> {
+    let len = buf.len();
+
+    let mut cur = Cursor::new(buf);
+    let mut holder = CodeHolder {
+        instructions: Vec::new(),
+        constant_pool: Vec::new(),
+    };
+
+    // check if this is a rvm bytecode file
+    // 52564D88
+    if !(cur.read_u32::<BigEndian>()? == 0x52564D88) {
```

Comment:
Same thing as my earlier comment, make the 0x52564D88 a constant: https://doc.rust-lang.org/rust-by-example/custom_types/constants.html